test/e2e/nodepool_upgrade_test.go

```diff
@@ -0,0 +1,114 @@
+	}
+
+	// Look up metadata about the release images so that we can extract the version
+	// information for later assertions
+	releaseInfoProvider := &releaseinfo.RegistryClientProvider{}
+	pullSecretFile, err := os.Open(clusterOpts.PullSecretFile)
+	g.Expect(err).NotTo(HaveOccurred(), "failed to open pull secret file")
+	defer pullSecretFile.Close()
+	pullSecret, err := ioutil.ReadAll(pullSecretFile)
+	g.Expect(err).NotTo(HaveOccurred(), "failed to read pull secret file")
+	previousReleaseInfo, err := releaseInfoProvider.Lookup(ctx, globalOpts.PreviousReleaseImage, pullSecret)
+	g.Expect(err).NotTo(HaveOccurred(), "failed to get release info for previous image")
+	latestReleaseInfo, err := releaseInfoProvider.Lookup(ctx, globalOpts.LatestReleaseImage, pullSecret)
+	g.Expect(err).NotTo(HaveOccurred(), "failed to get release info for latest image")
+
+	// Create the test cluster
+	hostedCluster := e2eutil.CreateCluster(t, ctx, client, &clusterOpts, hyperv1.AWSPlatform, globalOpts.ArtifactDir)
+
+	// Wait for connectivity to the cluster
+	t.Logf("Waiting for guest client to become available")
+	guestClient := e2eutil.WaitForGuestClient(t, ctx, client, hostedCluster)
+
+	// Wait for Nodes to be Ready
+	numNodes := int32(globalOpts.configurableClusterOptions.NodePoolReplicas * len(clusterOpts.AWSPlatform.Zones))
+	e2eutil.WaitForNReadyNodes(t, ctx, guestClient, numNodes)
+
+	// Wait for the first rollout to be complete and refresh the hostedcluster
+	t.Logf("Waiting for initial cluster rollout. Image: %s", hostedCluster.Spec.Release.Image)
+	e2eutil.WaitForImageRollout(t, ctx, client, hostedCluster, hostedCluster.Spec.Release.Image)
+	err = client.Get(ctx, crclient.ObjectKeyFromObject(hostedCluster), hostedCluster)
+	g.Expect(err).NotTo(HaveOccurred(), "failed to get hostedcluster")
+
+	// Find nodepools
+	var nodePools hyperv1.NodePoolList
+	err = client.List(ctx, &nodePools, &crclient.ListOptions{Namespace: hostedCluster.Namespace})
+	g.Expect(err).NotTo(HaveOccurred(), "failed to list nodepools")
+
+	// Wait for nodepools to roll out the initial version
+	// TODO: Consider doing this in parallel
+	for _, nodePool := range nodePools.Items {
+		e2eutil.WaitForNodePoolVersion(t, ctx, client, &nodePool, previousReleaseInfo.Version())
+	}
+
+	// Update nodepool images to the latest
+	for _, nodePool := range nodePools.Items {
+		err = client.Get(ctx, crclient.ObjectKeyFromObject(&nodePool), &nodePool)
+		g.Expect(err).NotTo(HaveOccurred(), "failed to get nodepool")
+		t.Logf("Updating nodepool image. Image: %s", globalOpts.LatestReleaseImage)
+		nodePool.Spec.Release.Image = globalOpts.LatestReleaseImage
+		err = client.Update(ctx, &nodePool)
+		g.Expect(err).NotTo(HaveOccurred(), "failed update nodePool image")
+	}
+
+	// Wait for Nodes to be Ready
+	e2eutil.WaitForNReadyNodes(t, ctx, guestClient, numNodes)
+
+	// Wait for nodepools to roll out the latest version
+	// TODO: Consider doing this in parallel
+	for _, nodePool := range nodePools.Items {
```

Comment:
This validates that the NodePool reflects the intended version in status.
Can we also validate the Nodes changed towards the targeted version by checking
```
  nodeInfo:
    architecture: amd64
    bootID: 4f51abfd-7c3c-48a2-b2ce-157a3a4d28d7
    containerRuntimeVersion: cri-o://1.23.0-107.rhaos4.10.git562c4e3.el8
    kernelVersion: 4.18.0-305.34.2.el8_4.x86_64
    kubeProxyVersion: v1.23.3+b63be7f
    kubeletVersion: v1.23.3+b63be7f
    machineID: ec2e9f9eaa80a8f58c3af2357200a83c
    operatingSystem: linux
    osImage: Red Hat Enterprise Linux CoreOS 410.84.202201302203-0 (Ootpa)
    systemUUID: ec2e9f9e-aa80-a8f5-8c3a-f2357200a83c
```
before and after?

Reply:
With HostedCluster, I thought our intent was to abstract the underlying details of the control plane (e.g. operator status, rollout status through HostedControlPlane, etc.) from black box observers like e2e which should only be consuming HostedCluster status. Is that also the case with NodePool? I'd expect NodePool status to report having rolled out a version only after all the underlying abstracted states have converged. Currently this seems to involve only the MachineDeployment, and not Nodes within the the control plane's worldview as you're describing.

Early on, before we had implemented those status elements with HostedCluster, e2e did reach in to the hosted control plane to inspect operator states directly, but that was considered temporary and was removed because only the control plane operator should encapsulate and surface those details. How do you think we should proceed with NodePool in this regard?

Reply:
>So I see checking the Node as a critical part of the expected output rather than as once of the black boxes.
>As a Hypershift engineer I can't claim with confidence that NodePool upgrades works without that check in CI, and to me being able to make that claim is ultimately the most important goal of the CI e2e testing.

I think we're saying the same thing here... the question is, who is responsible for the Node check? Seems like it would be the nodepool controller itself? If the Node check is essential, then the nodepool controller seems deficient without incorporating Nodes into its NodePool status calculations

Reply:
ah right I see what you mean! The NodePool controller today trusts the machineDeployment signal which programatically seems fair and the right boundary to me, e.g each piece trust its contract and then we e2e test input/output. However I think your arg is good and valid and it'd be reasonable to let the NodePool controller to this check as well. But even if we do that I'd expect the e2e to do it as well because the Node version it's part of the expected outcome to me, we can't assume the NodePool controller is doing what we want it to do.

Reply:
Upon closer inspection... what are we comparing the node info to? None of that information is present on a nodepool itself, or in `/release-manifests/0000_50_installer_coreos-bootimages.yaml` from the release image — is there metadata elsewhere we should be extracting to use for verification?